sway-lib-std/src/storage/storage_api.sw

```diff
@@ -137,11 +133,54 @@ pub fn read<T>(slot: b256, offset: u64) -> Option<T> {
 /// }
 /// ```
 #[storage(write)]
-pub fn clear<T>(slot: b256) -> bool {
-    // Get the number of storage slots needed based on the size of `T` as the ceiling of 
-    // `__size_of::<T>() / 32`
-    let number_of_slots = (__size_of::<T>() + 31) >> 5;
+pub fn clear<T>(slot: b256, offset: u64) -> bool {
+    if __size_of::<T>() == 0 {
+        return true;
+    }
+
+    // Determine how many slots and where the value is to be cleared.
+    let (offset_slot, number_of_slots, _place_in_slot) = slot_calculator::<T>(slot, offset);
 
     // Clear `number_of_slots * 32` bytes starting at storage slot `slot`.
-    __state_clear(slot, number_of_slots)
+    __state_clear(offset_slot, number_of_slots)
+}
+
+/// Given a slot, offset, and type this function determines where something should be stored.
+///
+/// # Arguments
+///
+/// * `slot`: [b256] - The starting address at which something should be stored.
+/// * `offset`: [u64] - The offset from `slot` to store the value.
+///
+/// # Returns
+///
+/// [b256] - The calculated offset slot to store the value.
+/// [u64] - The number of slots the value will occupy in storage.
+/// [u64] - The word in the slot where the value will start.
+fn slot_calculator<T>(slot: b256, offset: u64) -> (b256, u64, u64) {
+    let size_of_t = __size_of::<T>();
+
+    // Get the last storage slot needed based on the size of `T`.
+    // ((offset * bytes_in_word) + bytes + (bytes_in_slot - 1)) >> align_to_slot = last slot
+    let last_slot = ((offset * 8) + size_of_t + 31) >> 5;
+
+    // Where in the storage slot to align `T` in order to pack word-aligned.
+    // offset % number_words_in_slot = word_place_in_slot
+    let place_in_slot = offset % 4;
+ 
+    // Get the number of slots `T` spans based on it's packed position.
+    // ((place_in_slot * bytes_in_word) + bytes + (bytes_in_slot - 1)) >> align_to_slot = number_of_slots
+    let number_of_slots = match __is_reference_type::<T>() {
+        true => ((place_in_slot * 8) + size_of_t + 31) >> 5,
+        false => 1,
+    };
+
+    // TODO: Update when u256 <-> b256 conversions exist.
+    // Determine which starting slot `T` will be stored based on the offset.
+    let mut u256_slot = asm(r1: slot) {r1: u256};
+    let u256_increment = asm(r1: (0, 0, 0, last_slot - number_of_slots)) { r1: u256 };
+    u256_slot += u256_increment;
+    let offset_slot = asm(r1: u256_slot) { r1: b256 };
```

Comment:
Why is the `u256_slot` needed here?
What happens if you create the `u256_increment` only?

Reply:
`u256_slot` is just the `b256` converted to a `u256`. This can be anything like: `0xkjbkjv4j5g6jhgchxhgjgkfdrz4fedz7xkj5h6gh4fd3xgf5624gxfhj45gcf23ds5z6g23h453fd5xf2d345chjkv2gc3g5f6dch3vkj4rhbwlhefkbasjetfwyjetrych2oi4h2c4oi2hn387489rg163bc4rg1934r8ncgn91348r7gc9134m8rg7cwegfkejnvasdfgmcowgri384g7cocx7g4rn98wgq7ccn8g374r93n84gnrceuefcfwf`.

`u256_increment` is the `u64` number of offsets, such as `10`, to increase the value above by converted to a `u256`.